internal/stages_test.go

```diff
@@ -1,157 +1,53 @@
-	}
-}
-
-func TestWriteTree(t *testing.T) {
-	m := NewStdIOMocker()
-	m.Start()
-	defer m.End()
-
-	exitCode := runCLIStage("write_tree", "test_helpers/stages/read_tree")
-	if !assert.Equal(t, 1, exitCode) {
-		failWithMockerOutput(t, m)
-	}
-	assert.Contains(t, m.ReadStdout(), "Expected")
-	assert.Contains(t, m.ReadStdout(), "Test failed")
-
-	m.Reset()
-
-	exitCode = runCLIStage("write_tree", "test_helpers/stages/write_tree")
-	if !assert.Equal(t, 0, exitCode) {
-		failWithMockerOutput(t, m)
-	}
-}
-
-func TestCreateCommit(t *testing.T) {
-	m := NewStdIOMocker()
-	m.Start()
-	defer m.End()
-
-	exitCode := runCLIStage("create_commit", "test_helpers/stages/write_tree")
-	if !assert.Equal(t, 1, exitCode) {
-		failWithMockerOutput(t, m)
-	}
-	assert.Contains(t, m.ReadStdout(), "Expected")
-	assert.Contains(t, m.ReadStdout(), "Test failed")
-
-	m.Reset()
-
-	t.Skip("Not done yet")
-
-	exitCode = runCLIStage("create_commit", "test_helpers/stages/create_commit")
-	if !assert.Equal(t, 0, exitCode) {
-		failWithMockerOutput(t, m)
-	}
-}
-
-func runCLIStage(slug string, dir string) (exitCode int) {
-	cwd, err := os.Getwd()
-	if err != nil {
-		panic(err)
-	}
-	return RunCLI(map[string]string{
-		"CODECRAFTERS_CURRENT_STAGE_SLUG": slug,
-		"CODECRAFTERS_SUBMISSION_DIR":     path.Join(cwd, dir),
-		"CODECRAFTERS_COURSE_PAGE_URL":    "test",
-	})
-}
-
-func failWithMockerOutput(t *testing.T, m *IOMocker) {
-	m.End()
-	t.Error(fmt.Sprintf("stdout: \n%s\n\nstderr: \n%s", m.ReadStdout(), m.ReadStderr()))
-	t.FailNow()
+func TestStages(t *testing.T) {
+	testCases := map[string]tester_utils_testing.TesterOutputTestCase{
+		"init_success": {
+			UntilStageSlug: "init",
+			CodePath: "./test_helpers/stages/init",
+			ExpectedExitCode: 0,
+			StdoutFixturePath: "./test_helpers/fixtures/init",
+			NormalizeOutputFunc: normalizeTesterOutput,
+		},
+		"read_blob_success": {
+			UntilStageSlug: "read_blob",
+			CodePath: "./test_helpers/stages/read_blob",
+			ExpectedExitCode: 0,
+			StdoutFixturePath: "./test_helpers/fixtures/read_blob",
+			NormalizeOutputFunc: normalizeTesterOutput,
+		},
+		"create_blob_success": {
+			UntilStageSlug: "create_blob",
+			CodePath: "./test_helpers/stages/create_blob",
+			ExpectedExitCode: 0,
+			StdoutFixturePath: "./test_helpers/fixtures/create_blob",
+			NormalizeOutputFunc: normalizeTesterOutput,
+		},
+		"read_tree_success": {
+			UntilStageSlug: "read_tree",
+			CodePath: "./test_helpers/stages/read_tree",
+			ExpectedExitCode: 0,
+			StdoutFixturePath: "./test_helpers/fixtures/read_tree",
+			NormalizeOutputFunc: normalizeTesterOutput,
+		},
+		"write_tree_success": {
+			UntilStageSlug: "write_tree",
+			CodePath: "./test_helpers/stages/write_tree",
+			ExpectedExitCode: 0,
+			StdoutFixturePath: "./test_helpers/fixtures/write_tree",
+			NormalizeOutputFunc: normalizeTesterOutput,
+		},
+	}
+
+	tester_utils_testing.TestTesterOutput(t, testerDefinition, testCases)
+}
+
+func normalizeTesterOutput(testerOutput []byte) []byte {
```

Comment:
it seems that the `CompareOutputWithFixture` function in tester utils requires a `NormalizeOutputFunc`. is it better that this is always required? not sure how to normalize output for git, so i just returned the output

Reply:
Yep, it's okay to just return the output if that works - some challenges will have randomness involved (like with TCP port numbers), that's where the normalization function is useful.